refactor!: Switch to the annotation-based validator

aababilov opened the pull request:
One of the primary goals is to reduce the amount of boilerplate code as much as possible. The following things should be doable with minimal amount of code (one or several lines).
1. Add a new field to an existing table and make sure that this field is read from CSV and undergoes basic validation for the datatype (integer, float, enum etc.)
1. Add a typical validation rule such as referential integrity for a foreign key, uniqueness for a primary key, checks for positive values for an integer etc.
1. Install a new validation rule.

The new validator takes advantage of multithreading.
1. Different tables can be read in parallel.
1. Most validity checks do not depend on each other, so they also should run in parallel.
1. Validation that depends only on table A should start when this table is read. It doesn't have to wait for other tables.

Here is how the schema looks for the routes.txt table.
 
```
@GtfsTable("routes.txt")
public interface GtfsRouteSchema extends GtfsEntity {
    @FieldType(FieldTypeEnum.ID)
    @PrimaryKey
    @Required
    String routeId();

    @FieldType(FieldTypeEnum.ID)
    @ForeignKey(table = "agency.txt", field = "agency_id")
    String agencyId();

    String routeShortName();

    String routeLongName();

    String routeDesc();

    GtfsRouteType routeType();

    @FieldType(FieldTypeEnum.URL)
    String routeUrl();

    @DefaultValue("FFFFFF")
    GtfsColor routeColor();

    @DefaultValue("000000")
    GtfsColor routeTextColor();

    @NonNegative
    int routeSortOrder();

    @DefaultValue("1")
    GtfsContinuousPickupDropOff continuousPickup();

    @DefaultValue("1")
    GtfsContinuousPickupDropOff continuousDropOff();
}
```

1. `@GtfsTable` annotation makes the parser aware of a table with a given name and structure.
1. Each method of the interface defines a field with a corresponding name in the GTFS table.
1. Field types are deduced from the return types of the methods.
1. More precise GTFS data types are specified with the `@FieldType` annotation.
1. `@PrimaryKey` annotation creates an index based on the given field and adds validation that field values must be unique.
1. `@Required` adds validation that a value must be provided for the field in the CSV.
1. `@ForeignKey` annotation adds a validation of referential integrity.
1. `@DefaultValue` annotation specifies the default value for a field.
1. `@NonNegative` annotation adds a validation that a field must hold a positive value.
1. `GtfsColor` is a special data type defined in the validator core.
1. `GtfsContinuousPickupDropOff` is an enum generated based on enum annotations (see below).

Annotation processor generates three classes to read and retain the table. It also generates classes to check referential integrity.

Sample command-line parameters for the OpenOV GTFS dataset from the Netherlands:

`-i /Users/myUserName/Downloads/gtfs.zip --threads 4 -o output -f nl-open`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MobilityData/gtfs-validator?pullRequest=486) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MobilityData/gtfs-validator?pullRequest=486) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MobilityData/gtfs-validator?pullRequest=486) it.</sub>

aababilov commented:
Task for the already existing :adapter:repository:in-memory-simple:test failed on cloud but passes locally.

fabrice-v commented:
> Task for the already existing :adapter:repository:in-memory-simple:test failed on cloud but passes locally.

Yeah we need to remove the memory consumption tests like I did here
https://github.com/MobilityData/gtfs-validator/pull/487/commits/37df4d3d7b8838ceb450eab754535900c77d4578

fabrice-v commented:
Branch is now passing all CI/CD, as can be seen with the Docker image available [here](https://github.com/orgs/MobilityData/packages/container/gtfs-validator/335158).
I'll let everybody involved toake over from now. Thks for all the great work.

aababilov commented:
> Yeah we need to remove the memory consumption tests like I did here
> [37df4d3](https://github.com/MobilityData/gtfs-validator/commit/37df4d3d7b8838ceb450eab754535900c77d4578)
> 
> @lionel-nj

Great, thanks!

barbeau commented:
@aababilov Thanks! I've started reviewing on my local machine and it looks good so far. Could you please go ahead and push the remaining 2 modules that you implemented to this PR, as well as remove all the current models? So maybe 3 more commits total (one for each of the above)? I'd like to merge all changes in one PR so it's a clean transition from one architecture to the other. It will also be easier for me to review as a whole if all new files are in the same branch.

aababilov commented:
> @aababilov Thanks! I've started reviewing on my local machine and it looks good so far. Could you please go ahead and push the remaining 2 modules that you implemented to this PR, as well as remove all the current models? So maybe 3 more commits total (one for each of the above)? I'd like to merge all changes in one PR so it's a clean transition from one architecture to the other. It will also be easier for me to review as a whole if all new files are in the same branch.

Thank you!

I added all my new code and removed the old one. I added documentation to all the new classes but the unit tests are still missing. I will add them in the next commits.

fabrice-v commented on application/web-app/react-client/src/App.js:
```diff
@@ -1,93 +0,0 @@
-/*
```
@barbeau I defer to you defining when and how we could migrate the local web UI. Maybe that module could be preserved in this PR? Maybe the whole `web-app` and subfolders structure could be kept? @aababilov

barbeau replied:
In the interest of getting the PR merged to master quickly, I'd say we merge this PR without it and then bring it back via a new PR. That way everyone can continue to iterate on the master branch.

fabrice-v replied:
Fine with me. Just know the branch is broken because of it. Should be an easy fix @lionel-nj

lionel-nj replied:
Just seeing this now! Thanks @fabrice-v for pointing out. I removed the jobs that relied on the existence of the local web ui()
I also added instructions to package the app with Gradle (maybe I should have just removed the workflow jobs that was triggered? I can still remove these lines of code if there is need). This led to also remove the `javadoc` related workflow jobs since they were also failing. I can always reintegrate them.

fabrice-v commented on core/build.gradle:
```diff
@@ -0,0 +1,22 @@
+    implementation 'com.google.guava:guava:29.0-jre'
+    implementation 'commons-validator:commons-validator:1.6'
+    testImplementation group: 'junit', name: 'junit', version: '4.13'
+    testImplementation "com.google.truth:truth:1.0.1"
```
Google truth hey? :D

barbeau replied:
I've seen Truth used in another project and it looks cool :)

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/FieldTypeEnum.java:
```diff
@@ -0,0 +1,21 @@
+
+public enum FieldTypeEnum {
+    INTEGER,
+    FLOAT,
```
In the spec under "Field Types" there is a "non-negative float" type - do you think that should be explicitly defined here?

barbeau replied:
I see later that you test for these types for this and INTEGER - what you have LGTM

fabrice-v replied:
Yes, this non negative guy is weird and we bumped into it repeateadly.

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/FieldTypeEnum.java:
```diff
@@ -0,0 +1,21 @@
+package org.mobilitydata.gtfsvalidator.annotation;
+
+public enum FieldTypeEnum {
+    INTEGER,
```
In the spec under "Field Types" there is a "non-negative integer" type - do you think that should be explicitly defined here?

asvechnikov2 replied:
Would it make sense to add a comment that sets up expectation about how to use INTEGER or FLOAT with regards to negative, non-negative, other potential properties? Also, maybe it makes sense to clarify how these types are related to the official GTFS specification

barbeau commented on processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/Analyser.java:
```diff
@@ -0,0 +1,144 @@
+                    case SHORT:
+                    case INT:
+                    case LONG:
+                    case CHAR:
```
Is it correct to have `CHAR` here? It seems out of place with the other fields to return as `INTEGER`.

barbeau commented on processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EntityImplementationGenerator.java:
```diff
@@ -0,0 +1,250 @@
+                case TEXT:
+                    return CodeBlock.of("$S", valueString);
+                default:
+                    // TODO: Support all types or throw an exception.
```
Just flagging the TODO here - we should implement in this PR or open an issue to finish it later.

barbeau replied:
Flagging for follow-up

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:
```diff
@@ -0,0 +1,360 @@
+        }
+    };
+    private final ValueParser<Locale> languageCodeParser = new ValueParser("language code") {
+        // TODO: Enhance checks for IETF BCP 47 language code.
```
Just flagging the TODO here - we should implement in this PR or open an issue to finish it later.

fabrice-v replied:
I propose to use the FIXME "code" instead of TODO.

aababilov replied:
Done.

barbeau commented on processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/Analyser.java:
```diff
@@ -0,0 +1,144 @@
+package org.mobilitydata.gtfsvalidator.processor;
```
We need the Apache license header on all classes with copyright Google 2020.

aababilov replied:
Done.

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsUnarchivedInput.java:
```diff
@@ -0,0 +1,36 @@
+                filenames.add(file.getName());
+            }
+        }
+
```
Nit - extra new line

aababilov replied:
Removed. I wonder why "Reformat code" in IDEA does not drop such redundant lines.

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/CsvFile.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    /**
+     * Tells if the file is empty, i.e. it has no rows and even no headers.
```
Recommend using normal Javadoc formatting with `@return`

aababilov replied:
Done.

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/CsvFile.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    /**
+     * Returns the next record or null if not available.
```
Recommend using normal Javadoc formatting with `@return`

aababilov replied:
Done.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/cli/Arguments.java:
```diff
@@ -0,0 +1,18 @@
+ */
+public class Arguments {
+
+    @Parameter(names = {"-i", "--input"}, description = "Location of the unarchived directory with input GTFS files", required = true)
```
I provided the file name `gtfs.zip` as the `-i` parameter and that worked. Can you provide a base directory that contains multiple zip files too? Suggest clarifying the description of this parameter.

aababilov replied:
This was an outdated comment. Fixed.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/notice/UnusedShapeNotice.java:
```diff
@@ -0,0 +1,16 @@
+package org.mobilitydata.gtfsvalidator.notice;
```
What are your thoughts on moving the `notice` package to the `core` module? I think it would be useful to have all the notice definitions `core` package so it could be used as a dependency in another project. Same with `validator` package that's currently in `main`. `main` package could be reduced to the CLI.

aababilov replied:
There is already a `notice` package in `core`. It contains the most basic notices such as `EmptyFileNotice`, `ForeignKeyError` that do not depend on particular schema.

Let's discuss in another comment whether to move the majority of `main` to `core`.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsAttributionSchema.java:
```diff
@@ -0,0 +1,39 @@
+package org.mobilitydata.gtfsvalidator.table;
```
So looking at this more given that the `main` depends on `processor` to auto-generate code, I guess it's debatable how useful the table interfaces are without the processor.

aababilov replied:
Unfortunately, this will lead to a circular dependency:
* `processor` depends on classes from `core` (`GtfsTable`, `ForeignKey`, `ForeignKeyError`)
* `core` depends on `processor` to generate classes from GTFS schema.

Here is what Gradle says:

```
:core:classes
\--- :core:compileJava
     +--- :core:jar
     |    \--- :core:classes (*)
     \--- :processor:jar
          \--- :processor:classes
               \--- :processor:compileJava
                    \--- :core:jar (*)

(*) - details omitted (listed previously)
```

> I think it would be useful to have the core GTFS schema definition in the core package so it could be used as a dependency in another project.

I can split `main` in two parts:
* `schemastd` - definition of the standard GTFS schema + specific validations;
* `cli` - the `org.mobilitydata.gtfsvalidator.cli` package 

What do you think?

barbeau replied:
>I can split main in two parts:
>
>* schemastd - definition of the standard GTFS schema + specific validations;
* cli - the org.mobilitydata.gtfsvalidator.cli package
>
>What do you think?

After looking at this further I could go either way on this. 

In the interest of development simplicity I think fewer modules are better. And, looking at it from a dependency-perspective, I don't think the CLI dependencies would really add much to a JAR. 

Alternately, if we want to have a strict separation between "application"-level classes and the data model it could make sense to split them.

@aababilov You have to the best handle on the code as-is so I'd defer to your opinion.

aababilov replied:
I agree, the CLI is currently so tiny right now that it does not make the difference. Let's stay with the current `main` subproject and we will split it later if needed.

barbeau reviewed (commented):
@aababilov Thanks for putting this together, it looks good! I still have more review to do but I figured I'd post what I have so far and then pick back up next week. Please see the comments in-line.

It's fast! It ran through the Netherlands dataset (~255MB with 12182720 records in stop_times.txt) in about 30 seconds with 8 threads on my machine. 🚀

f8full commented on core/src/test/java/org/mobilitydata/gtfsvalidator/type/GtfsTimeTest.java:
```diff
@@ -0,0 +1,21 @@
+public class GtfsTimeTest {
+    @Test
+    public void fromString() {
+        assertThat(GtfsTime.fromString("12:20:30").getSecondsSinceMidnight()).isEqualTo(12 * 3600 + 20 * 60 + 30);
```
Reading the conversation, may I suggest looking at our current tests and try to migrate them if there are any that already cover what you identify as missing?

barbeau replied:
@f8full Agreed - here are the current time tests - https://github.com/MobilityData/gtfs-validator/blob/master/adapter/utils/src/test/java/org/mobilitydata/gtfsvalidator/timeutils/TimeUtilsImplTest.java#L24

aababilov commented:
> @aababilov Thanks for putting this together, it looks good! I still have more review to do but I figured I'd post what I have so far and then pick back up next week. Please see the comments in-line.
> 
> It's fast! It ran through the Netherlands dataset (~255MB with 12182720 records in stop_times.txt) in about 30 seconds with 8 threads on my machine.

aababilov commented on core/build.gradle:
```diff
@@ -0,0 +1,22 @@
+    implementation 'com.univocity:univocity-parsers:2.9.0'
+    implementation 'com.fasterxml.jackson.core:jackson-databind:2.10.0'
+    implementation 'com.google.guava:guava:29.0-jre'
+    implementation 'commons-validator:commons-validator:1.6'
```
I use `org.apache.commons.validator.routines.EmailValidator` and `org.apache.commons.validator.routines.UrlValidator` from this library.

aababilov commented on core/src/test/java/org/mobilitydata/gtfsvalidator/parsing/RowParserTest.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    @Test
+    public void asPhoneNumber() {
```
Well, parsing phone numbers is a hard task! Validity of the phone number depends on country. So, I have to add a flag to specify a feed name (such as `nl-openov`) and deduct country code from it.

aababilov commented on core/src/test/java/org/mobilitydata/gtfsvalidator/parsing/RowParserTest.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Test
+    public void asTime() {
+        assertThat(createParser("12:20:30").asTime(0, true)).isEqualTo(GtfsTime.fromHourMinuteSecond(12, 20, 30));
```
Done,

aababilov commented on core/src/test/java/org/mobilitydata/gtfsvalidator/parsing/RowParserTest.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Test
+    public void asLatitude() {
+        assertThat(createParser("32.5").asLatitude(0, true)).isEqualTo(32.5);
```
Done.

asvechnikov2 commented on core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/FieldType.java:
```diff
@@ -0,0 +1,15 @@
+import java.lang.annotation.Target;
+
+/**
+ * Specifies type of a GTFS field, e.g., {@code COLOR} or {@code LATITUDE}.
```
What do you think about saying "{@code COLOR}, {@code LATITUDE}, etc", to explicitly show that there are more than just two field types? Also, could you add an expectation of how this annotation is going to be used? In other words, as a developer I'd like to know answer to questions such as
 * Do I have to specify this annotation?
 * What benefits will I get by specifying this notification?
 * What Java data types must I use with a particular FieldType (if any)? I think, FieldTypeEnum should clarify this per field type and here we should say that there's such limitation are applied. Just to add a bit more context, what can I use with FieldTypeEnum.INTEGER? 'short', 'int', 'long', BigInteger?

aababilov replied:
Very good point. I added documentation.

I say "e.g, COLOR or LATITUDE", so it suggests that I am giving examples and not a full list :)

asvechnikov2 commented on core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/Generated.java:
```diff
@@ -0,0 +1,15 @@
+import static java.lang.annotation.RetentionPolicy.SOURCE;
+
+/**
+ * Marker for all classes generated by annotation processor.
```
I'm yet to review other parts, but my understanding is that this annotation is only for intermediate usage during code generation. Should we explicitly mention that this annotation isn't expected to be used in the codebase?

also, nit: this file is the only file that uses static import for TYPE and SOURCE, probably you want to change it to usual import to be consistent with other classes.

aababilov replied:
Comment added!

I just rely on the automatic cleanups for imports in IDEA. It would be a C++ nightmare to maintain all imports manually :)

asvechnikov2 commented on core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/GtfsEnumValue.java:
```diff
@@ -0,0 +1,12 @@
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.SOURCE)
+@Repeatable(GtfsEnumValues.class)
+public @interface GtfsEnumValue {
```
/**
 \* Annotation to specify an enum name-value pair. The pair will be used to generate an actual Enum that will be used by the processing code.
 */

aababilov replied:
Done.

asvechnikov2 commented on core/src/main/java/org/mobilitydata/gtfsvalidator/annotation/GtfsEnumValues.java:
```diff
@@ -0,0 +1,12 @@
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.SOURCE)
+public @interface GtfsEnumValues {
```
Could you please add documentation explaining how this annotation must be used if it can be used and what's the difference with GtfsEnumValue?

aababilov replied:
It must not be used. Added documentation.

asvechnikov2 reviewed (commented):
Posting first comments for 'gtfsvalidator.annotation.*' . Could you please add docstrings for all annotations to setup expectations for developers on the use of these annotations? This will be very helpful for future developers as well as help with the reviewing process.

fabrice-v commented on .circleci/config.yml:
```diff
@@ -10,7 +10,6 @@ jobs:
           name: Package cli jar with Gradle
           command: |
             ./gradlew shadowJar
-
```
👍

asvechnikov2 reviewed (commented):
Comments for "Add the core part of the annotation-based validator" commit

fabrice-v commented on core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:
```diff
@@ -0,0 +1,360 @@
+        }
+    }
+
+    public String asString(int columnIndex, boolean required) {
```
And if I may add, in the current implementation we finalized all parameters (and actually anything and everything we could). Shall that be done here too?

aababilov replied:
I am a big fan of const variables but the Google Java guide discourages the general practice of marking all parameters and local variables as final (without giving a rationale, though).

I am fine with making everything final if most of us agree.

aababilov replied:
I added `final` here and for fields in other classes.

fabrice-v commented on core/src/main/java/org/mobilitydata/gtfsvalidator/type/GtfsTime.java:
```diff
@@ -0,0 +1,92 @@
+
+    public static GtfsTime fromString(String hhmmss) {
+        int length = hhmmss.length();
+        if (length != 8 && length != 7) {
```
Dang! We completely missed that in the current implementation. It also requires some spec rewrite

>Time in the HH:MM:SS format (H:MM:SS is also accepted)

(cc @timMillet ).

I am not sure how the Google python validator handle such cases. Thks for pointing it out!

aababilov replied:
Google allows up to 15 digits for hours but we don't support journeys that last for several days on practice. Let's allow up to 3 digits for hours.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsInput.java:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Lists all files inside the GTFS dataset, even if they are not CSV and do not have .txt extension.
+     *
+     * @return names of all available files
```
Done. Covered with tests.

aababilov commented on core/src/test/java/org/mobilitydata/gtfsvalidator/input/GtfsInputTest.java:
```diff
@@ -0,0 +1,53 @@
+import static org.junit.Assert.assertThrows;
+
+@RunWith(JUnit4.class)
+public class GtfsInputTest {
```
Done.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/CsvFile.java:
```diff
@@ -0,0 +1,120 @@
+     * Returns the next record or null if not available.
+     */
+    @Nullable
+    public CsvRow nextResult() {
```
I made nextResult() private and documented iterator().

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/RowParser.java:
```diff
@@ -0,0 +1,360 @@
+                case NON_ZERO:
+                    if (value == 0) {
+                        parseErrorsInRow = true;
+                        noticeContainer.addNotice(new NumberOutOfBoundsError(
```
Done. Good point.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/validator/TableHeaderValidator.java:
```diff
@@ -0,0 +1,49 @@
+                            Set<String> supportedColumns, Set<String> requiredColumns,
+                            NoticeContainer noticeContainer) {
+        boolean isValid = true;
+        if (actualColumns == null) {
```
No, it does not. Otherwise CsvFileTest would not pass.

asvechnikov2 replied:
Sorry for the confusion. CsvFile always sets columnNames to an empty array if file is empty or columns are null. E.g. it should be impossible for actualColumns to be equal to null (that was my perception).

aababilov replied:
Wow, how did you find this bug? Thanks a lot! Fixed. I also added unit tests.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/validator/ValidatorLoader.java:
```diff
@@ -0,0 +1,190 @@
+import java.util.Map;
+
+/**
+ * A validator that handles a single entity and does not need more information.
```
Fixed.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/validator/ValidatorLoader.java:
```diff
@@ -0,0 +1,190 @@
+        for (Class<? extends SingleEntityValidator> validatorClass : singleEntityValidatorClasses) {
+            for (Method method : validatorClass.getMethods()) {
+                if (method.getName().equals("validate") && method.getParameterCount() == 2
+                        && GtfsEntity.class.isAssignableFrom(method.getParameterTypes()[0])
+                        && !method.getParameterTypes()[0].isAssignableFrom(GtfsEntity.class)
```
This is RTTI, it is not clear because programmer normally do not deal with it :) 

Documented.

fabrice-v replied:
Real Time Type Information? I did that in C++ back in the days. Felt like nitroglycerin. :D

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NoticeContainer.java:
```diff
@@ -0,0 +1,64 @@
+import java.util.Map;
+
+public class NoticeContainer {
+    private static int MAX_NOTICES_TO_STORE = 100000;
```
Done.

lionel-nj commented on core/build.gradle:
```diff
@@ -0,0 +1,23 @@
+group 'org.mobilitydata'
+version "gtfs-validator-${System.getenv("versionTag")}"
+
+sourceCompatibility = 1.8
```
Out of curiosity, we used to have 
```sourceCompatibility = JavaVersion.VERSION_11``` why was this change required?

aababilov replied:
This makes sure that we can run Validator at Google.

Google's products are one step ahead but the technologies may be outdated :(

lionel-nj replied:
Thanks!

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsAgencySchema.java:
```diff
@@ -0,0 +1,35 @@
+
+@GtfsTable("agency.txt")
+public interface GtfsAgencySchema extends GtfsEntity {
+    @FieldType(FieldTypeEnum.ID)
```
@aababilov What would you think of adding a `@ConditionallyRequired` annotation for fields like `agency_id` where GTFS has the field as "Conditionally Required"? Looks like this is currently just implemented in `AgencyConsistencyValidator` as a validation rule, but IMHO it would be nice to have some indicator on these fields that they are subject to special treatment (i.e., they aren't "Optional"). Even if the annotation isn't used for generating code I think the indicator would still be helpful for devs to check and make sure that current and future fields are being treated according to the spec.

aababilov replied:
Good point! Done.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsAgencySchema.java:
```diff
@@ -0,0 +1,35 @@
+    String agencyUrl();
+
+    @Required
+    TimeZone agencyTimezone();
```
Should this be `@FieldType(FieldTypeEnum.TIMEZONE)` as well?

aababilov replied:
Annotation processor deducts field type in most cases (documented in FieldTypeEnum.java):

 * Many GTFS types are deducted from the actual Java types in schema definition:
 * <p>
 * * {@code int} - {@code INTEGER};
 * * {@code double} - {@code FLOAT};
 * * {@code String} - {@code TEXT};
 * * {@code boolean} - {@code BOOLEAN};
 * * {@code GtfsColor} - {@code COLOR};
 * * {@code GtfsDate} - {@code DATE};
 * * {@code GtfsTime} - {@code TIME};
 * * {@code TimeZone} - {@code TIMEZONE};
 * * {@code Locale} - {@code LANGUAGE_CODE};
 * * {@code Currency} - {@code CURRENCY_CODE}.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsAgencySchema.java:
```diff
@@ -0,0 +1,35 @@
+    @Required
+    TimeZone agencyTimezone();
+
+    Locale agencyLang();
```
Should this be `@FieldType(FieldTypeEnum.LANGUAGE_CODE)` as well?

aababilov replied:
Ditto.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFareAttributeSchema.java:
```diff
@@ -0,0 +1,32 @@
+    String fareId();
+
+    @NonNegative
+    double price();
```
Related - there is currently a discussion at https://github.com/google/transit/issues/254 for potentially adding a new `decimal` semantic data type to GTFS for fares.

aababilov replied:
Done.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsAttributionSchema.java:
```diff
@@ -0,0 +1,39 @@
+    @ForeignKey(table = "routes.txt", field = "route_id")
+    String routeId();
+
+    @ForeignKey(table = "trips.txt", field = "trip_id")
```
Should also be `@FieldType(FieldTypeEnum.ID)`?

aababilov replied:
Done.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsAttributionSchema.java:
```diff
@@ -0,0 +1,39 @@
+    @Required
+    String organizationName();
+
+    boolean isProducer();
```
Should these "Enums" in GTFS data types be enums instead of boolean values? It's possible (although probably unlikely in this particular field case) that a `2` value could be added in the future. Looks like GTFS actually doesn't have a "Boolean" data type.

aababilov replied:
Right, I've spent a lot of time choosing between boolean and enum. I agree that enums are better because they are flexible.

barbeau reviewed (commented):
@aababilov Here's my next review chunk. Still looking good! 👍

lionel-nj commented on main/src/main/java/org/mobilitydata/gtfsvalidator/cli/Main.java:
```diff
@@ -0,0 +1,57 @@
+        System.out.println("Input: " + args.input);
+        System.out.println("Output: " + args.outputBase);
+        System.out.println("Table loaders: " + feedLoader.listTableLoaders());
+        System.out.println("Validators:\n" + validatorLoader.listValidators());
```
```suggestion
        System.out.println("Validators:" + System.lineSeparator() + validatorLoader.listValidators());
```

lionel-nj replied:
Since the line separator can vary from an OS to another one.

aababilov replied:
Done.

lionel-nj commented on main/src/main/java/org/mobilitydata/gtfsvalidator/notice/DecreasingShapeDistanceNotice.java:
```diff
@@ -0,0 +1,24 @@
+
+import com.google.common.collect.ImmutableMap;
+
+public class DecreasingShapeDistanceNotice extends Notice {
```
For consistency with the other notices, I suggest reming this one to `DecreasingShapeDistance`

lionel-nj replied:
Another solution would be to append the term "notice" to the other notices:
- `InconsistentAgencyField`
- `StartAndEndDateOutOfOrder`
- `StopTimeWithArrivalBeforePreviousDepartureTime`
- `StopTimeWithDepartureBeforeArrivalTime`
- `StopTimeWithOnlyArrivalOrDepartureTime`
- `UnusedShape`

aababilov replied:
Notices look similar to Java exceptions. Since exceptions have a suffix (such as child classes here: https://docs.oracle.com/javase/7/docs/api/java/lang/RuntimeException.html), I decided to add a suffix to notice classes.

lionel-nj commented on main/src/main/java/org/mobilitydata/gtfsvalidator/notice/DecreasingShapeDistanceNotice.java:
```diff
@@ -0,0 +1,24 @@
+
+import com.google.common.collect.ImmutableMap;
+
+public class DecreasingShapeDistanceNotice extends Notice {
```
```suggestion
public class DecreasingShapeDistance extends Notice {
```

lionel-nj commented on main/src/main/java/org/mobilitydata/gtfsvalidator/notice/RouteColorContrastNotice.java:
```diff
@@ -0,0 +1,18 @@
+import com.google.common.collect.ImmutableMap;
+import org.mobilitydata.gtfsvalidator.type.GtfsColor;
+
+public class RouteColorContrastNotice extends Notice {
```
For consistency with the other notices, I suggest reming this one to `RouteColorContrast`.

Another solution would be to append the term "notice" to the other notices' class name:
- `InconsistentAgencyField`
- `StartAndEndDateOutOfOrder`
- `StopTimeWithArrivalBeforePreviousDepartureTime`
- `StopTimeWithDepartureBeforeArrivalTime`
- `StopTimeWithOnlyArrivalOrDepartureTime`
- `UnusedShape`

lionel-nj replied:
```suggestion
public class RouteColorContrast extends Notice {
```

lionel-nj commented on main/src/main/java/org/mobilitydata/gtfsvalidator/notice/UnusedShapeNotice.java:
```diff
@@ -0,0 +1,16 @@
+
+import com.google.common.collect.ImmutableMap;
+
+public class UnusedShapeNotice extends Notice {
```
For consistency with the other notices, I suggest reming this one to `UnusedShape`.

Another solution would be to append the term "notice" to the other notices' class name:
- `InconsistentAgencyField`
- `StartAndEndDateOutOfOrder`
- `StopTimeWithArrivalBeforePreviousDepartureTime`
- `StopTimeWithDepartureBeforeArrivalTime`
- `StopTimeWithOnlyArrivalOrDepartureTime`
- `UnusedShape`

lionel-nj replied:
```suggestion
public class UnusedShape extends Notice {
```

aababilov commented on main/src/main/java/org/mobilitydata/gtfsvalidator/cli/Arguments.java:
```diff
@@ -0,0 +1,21 @@
+    @Parameter(names = {"-t", "--threads"}, description = "Number of threads to use")
+    public int numThreads = 1;
+
+    @Parameter(names = {"-f", "--feed_name"}, description = "Name of the feed, e.g., `nl-openov`")
```
Wow, is there an ISO format for feed names? Could you give me a link that I can paste here?

aababilov replied:
Oh, you mean, ISO 3166-1 alpha-2 for the country code, not for the whole feed name. Fixed.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,117 @@
+            try {
+                loader = clazz.asSubclass(GtfsTableLoader.class).getConstructor().newInstance();
+            } catch (ReflectiveOperationException exception) {
+                System.err.println("Possible bug in GTFS annotation processor");
```
This may happen if annotation processor did not generate a constructor without parametres.

lionel-nj commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFareAttributeSchema.java:
```diff
@@ -0,0 +1,32 @@
+    double price();
+
+    @Required
+    Currency currencyType();
```
Should this be annotated as "@FieldType(FieldTypeEnum.CURRENCY_CODE)"?

aababilov replied:
No because it is already a `Currency`.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,117 @@
+        ClassPath classPath;
+        try {
+            classPath = ClassPath.from(ClassLoader.getSystemClassLoader());
+        } catch (IOException exception) {
```
Done

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    private static Reader createFileReader(InputStream stream) throws FileNotFoundException {
```
Cool, removed.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public GtfsFeedContainer load(GtfsInput gtfsInput, GtfsFeedName feedName, ValidatorLoader validatorLoader,
+                                  NoticeContainer noticeContainer) throws IOException, InterruptedException {
```
Great, removed.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,117 @@
+            for (FileValidator validator : validatorLoader.createMultiFileValidators(feed)) {
+                validatorCallables.add(() -> {
+                    validator.validate(noticeContainer);
+                    return null;
```
The `load()` method does not return null and does not throw exceptions. Do you want me to throw all InterruptedException instead of catching and printing them?

barbeau replied:
Ah, right, sorry, I missed that. No, I think it's fine as-is.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,117 @@
+                noticeContainer.addNotice(new UnexpectedFile(filename));
+            } else {
+                loaderCallables.add(() -> loader.load(
+                        createFileReader(gtfsInput.getFile(filename)), feedName, validatorLoader, noticeContainer));
```
Oh, sure! Done.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsZipFileInput.java:
```diff
@@ -0,0 +1,55 @@
+        // UNIX file systems etc.
+        //
+        // Directory names in end with '/'.
+        return entry.getName().contains("/");
```
Done. However, I think that the existing comment was already clear.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsUnarchivedInput.java:
```diff
@@ -0,0 +1,35 @@
+
+    public GtfsUnarchivedInput(File directory) throws IOException {
+        this.directory = directory;
+        for (File file : directory.listFiles()) {
```
Let's keep it simple. A feed has about a dozen files, so DirectoryStream API does not add much value.

What really needs optimization is reading a CSV file. Reading and parsing takes much more time than further validations right now.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/notice/NoticeContainer.java:
```diff
@@ -0,0 +1,64 @@
+            for (Notice notice : noticesOfType) {
+                ++i;
+                if (i > MAX_EXPORTS_PER_NOTICE_TYPE) {
+                    // Do not store too many notices.
```
Done.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/parsing/CsvFile.java:
```diff
@@ -0,0 +1,132 @@
+    public CsvFile(Reader reader, String filename) {
+        this.filename = filename;
+        CsvParserSettings settings = new CsvParserSettings();
+        settings.getFormat().setLineSeparator("\n");
```
Oh, nice. I added this test.


By the way, RFC 4180 prescribes to use CRLF but [our reference](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md) allows LF as well.

aababilov commented on core/src/main/java/org/mobilitydata/gtfsvalidator/validator/TableHeaderValidator.java:
```diff
@@ -0,0 +1,49 @@
+            return isValid;
+        }
+        Map<String, Integer> columnIndices = new HashMap<>();
+        // Sorted tree set for stable order of notices.
```
I would like to keep the comment because the goal is to have a stable order of notices. Stable order of column names is just a measure.

asvechnikov2 commented on processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/TableLoaderGenerator.java:
```diff
@@ -0,0 +1,218 @@
+                        "validatorLoader.invokeSingleFileValidators(table, noticeContainer)")
+                .addStatement("return table")
+                .endControlFlow()
+                .beginControlFlow("if (!new $T().validate(FILENAME, csvFile.getColumnNames(), " +
+                                "getColumnNames(), getRequiredColumnNames(), noticeContainer))",
```
If I understand correctly, csvFile.getColumnNames() returns column names as they are written in CSV file. getColumnNames and getRequireColumnNames both return something like LONG_NAME_FIELD_NAME. Am I missing anything?

aababilov replied:
All these methods return lower_case_names.

```
public static final String LEVEL_ID_FIELD_NAME = "level_id";

  public static final String LEVEL_INDEX_FIELD_NAME = "level_index";

  public static final String LEVEL_NAME_FIELD_NAME = "level_name";
  @Override
  public Set<String> getColumnNames() {
    return ImmutableSet.of(LEVEL_ID_FIELD_NAME, LEVEL_INDEX_FIELD_NAME, LEVEL_NAME_FIELD_NAME);
  }

  @Override
  public Set<String> getRequiredColumnNames() {
    return ImmutableSet.of(LEVEL_ID_FIELD_NAME, LEVEL_INDEX_FIELD_NAME);
  }
```

aababilov commented on processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/EntityImplementationGenerator.java:
```diff
@@ -0,0 +1,250 @@
+                    return CodeBlock.of("$T.fromInt(0x$L)",
+                            GtfsColor.class, Integer.toHexString(Integer.parseInt(valueString, 16)));
+                case TEXT:
+                    return CodeBlock.of("$S", valueString);
```
This would be bug-prone since it is not a trivial task to escape an arbitrary string to a Java string literal.
$S is the placeholder that handles escaping. It will add the double quotes and other things.

lionel-nj commented on core/src/main/java/org/mobilitydata/gtfsvalidator/input/GtfsUnarchivedInput.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public Set<String> getFilenames() {
+        return filenames;
```
```suggestion
        return  Collections.unmodifiableSet(filenames);
```

aababilov replied:
Done.

fabrice-v commented:
@barbeau @lionel-nj Do we know why the `doc` check is forever like this?
![image](https://user-images.githubusercontent.com/55884852/100607200-96d81380-32d8-11eb-9414-72e192936978.png)

Maybe a good first step would be to remove it from the `required` checks list?

lionel-nj commented:
> @barbeau @lionel-nj Do we know why the doc check is forever like this?

Investigating this @fabrice-v

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsRouteSchema.java:
```diff
@@ -0,0 +1,61 @@
+
+    String routeDesc();
+
+    GtfsRouteType routeType();
```
Should be `@Required`

aababilov replied:
Done.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsRouteSchema.java:
```diff
@@ -0,0 +1,61 @@
+    @DefaultValue("FFFFFF")
+    GtfsColor routeColor();
+
+    GtfsColor routeTextColor();
```
Should this be `@DefaultValue("000000")`?

barbeau replied:
Flagging for follow-up

aababilov replied:
This was already fixed.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsRouteSchema.java:
```diff
@@ -0,0 +1,61 @@
+    GtfsColor routeTextColor();
+
+    @NonNegative
+    int sortOrder();
```
For consistency this should probably be named `routeSortOrder()`

aababilov replied:
Done.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsStopTimesTimepointEnum.java:
```diff
@@ -0,0 +1,24 @@
+
+import org.mobilitydata.gtfsvalidator.annotation.GtfsEnumValue;
+
+@GtfsEnumValue(name = "EXACT", value = 0)
```
This is inverted - `1` is `EXACT`

aababilov replied:
Nice catch! Fixed.

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsTransferSchema.java:
```diff
@@ -0,0 +1,38 @@
+    @ForeignKey(table = "stops.txt", field = "stop_id")
+    String toStopId();
+
+    GtfsTransferType transferType();
```
Should be `@Required`

aababilov replied:
Done.

barbeau commented on processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/FieldNameConverter.java:
```diff
@@ -0,0 +1,62 @@
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Collection of utility functions to generate names of getters and setters, fields and other related staff for a given
```
```suggestion
 * Collection of utility functions to generate names of getters and setters, fields and other related stuff for a given
```

aababilov replied:
I learned the words "staff" and "stuff" many years ago when I was unable to hear a difference between them. Today I can pronounce both words correctly but I cannot remember which one is which. Do you know a mnemonics? :)

fabrice-v replied:
I personally make the connection in my head with stuffing a Turkey, filling it with -yummy- stuff :)

barbeau commented on main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Override
+    public void validate(NoticeContainer noticeContainer) {
+        for (List<GtfsShape> shapeList : Multimaps.asMap(table.byShapeIdMap()).values()) {
```
These are sorted because of the `@SequenceKey` annotation on `shape_pt_sequence`, right? So this should work.

aababilov replied:
Yep, I added a comment to mention that.

barbeau reviewed (commented):
@aababilov Another review chunk

aababilov commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFareAttributeSchema.java:
```diff
@@ -0,0 +1,50 @@
+    @Required
+    String fareId();
+
+    @NonNegative
```
Done.

aababilov commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFareAttributeTransfersEnum.java:
```diff
@@ -0,0 +1,26 @@
+
+import org.mobilitydata.gtfsvalidator.annotation.GtfsEnumValue;
+
+@GtfsEnumValue(name = "NO_TRANSFER", value = 0)
```
This field is quite special in the spec ;) It is a "required that can be empty" and its "empty" value differs from all its numeric values. We may introduce the magical -1, as you suggested, or we can simply use the `hasTransfers()` method to tell if this field is empty. Both ways do not look elegant. Perhaps a magical -1 is less error prone but using `hasTransfers()` is less hackish. Which one do you prefer?

barbeau replied:
>This field is quite special in the spec ;)

Yes, I thought the same thing when I looked at it. It's the only field defined this way in GTFS from what I saw, which makes it very much a corner case. I'd err on the side of being less error prone in this case, so I'd prefer the magical -1 solution.

aababilov replied:
OK, I added this magical -1. However, this allows to specify that -1 explicitly in the CSV file - which would be against the GTFS spec. Are you OK with that?

barbeau replied:
>However, this allows to specify that -1 explicitly in the CSV file - which would be against the GTFS spec. Are you OK with that?

Ugh - no, we shouldn't allow that. We need to only support canonical GTFS values. Can we still flag -1 as invalid somehow? If it's not possible to only set `UNLIMITED` in the case where the field is empty (and still flag a `-1` value in the GTFS data as an error), then I believe our only choice as you mentioned previously is to rely on calling code to use `hasTransfers()`. In that case, we should add as much documentation as possible in the Javadocs to explain this behavior.

aababilov replied:
I restored the previous situation. This is the simplest way to achieve what we need: this field may be either empty or keep 0, 1 or 2, and this is validated automatically.

I can tell you that Google does not implement any specific checks based on concrete value of this field, so the risk that someone forgets to call `hasTransfers()` is very low.

Anyway, I would advocate for adding -1 (or another explicit value) to the GTFS standard and also marking this field as `Optional` instead of `Required` in the transit/reference.md.

barbeau replied:
@aababilov Sounds good, thanks! I'll pass on the comments re: spec changes to MobilityData spec team.

timMillet replied:
Unfortunately, this change would not be backward compatible, so impossible to pass under the current GTFS amendment process. However, this could move forward if GTFS Schedule was versioned. I added the issue with this field in our records!

aababilov commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedInfoSchema.java:
```diff
@@ -0,0 +1,50 @@
+    @Required
+    String feedPublisherName();
+
+    @FieldType(FieldTypeEnum.URL)
```
Done.

aababilov commented on main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedInfoSchema.java:
```diff
@@ -0,0 +1,50 @@
+    @FieldType(FieldTypeEnum.URL)
+    String feedPublisherUrl();
+
+    Locale feedLang();
```
Done.

lionel-nj commented:
`Semantic Pull Request` bot has been disabled. All checks should pass for the next commit.

barbeau commented on core/src/test/java/org/mobilitydata/gtfsvalidator/parsing/RowParserTest.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Test
+    public void asLongitude() {
+        assertThat(createParser("-32.5").asLongitude(0, true)).isEqualTo(-32.5);
```
Flagging for follow-up

aababilov replied:
Already done a week ago :)

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,116 @@
+        List<Callable<GtfsTableContainer>> loaderCallables = new ArrayList<>();
+        Map<String, GtfsTableLoader> remainingLoaders = (Map<String, GtfsTableLoader>) tableLoaders.clone();
+        for (String filename : gtfsInput.getFilenames()) {
+            GtfsTableLoader loader = remainingLoaders.remove(filename.toLowerCase());
```
Flagging for follow-up

aababilov replied:
I convert filename to lower case in annotation processor.

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,116 @@
+        try {
+            try {
+                List<Future<GtfsTableContainer>> results = exec.invokeAll(loaderCallables);
+                tableContainers.ensureCapacity(results.size());
```
Flagging for follow-up

aababilov replied:
Fixed.

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,116 @@
+        this.numThreads = numThreads;
+    }
+
+    public GtfsFeedContainer load(GtfsInput gtfsInput, ValidatorLoader validatorLoader,
```
Flagging for follow-up

aababilov replied:
Renamed.

barbeau commented on core/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFeedLoader.java:
```diff
@@ -0,0 +1,116 @@
+                    try {
+                        tableContainers.add(f.get());
+                    } catch (ExecutionException | InterruptedException e) {
+                        e.printStackTrace();
```
Flagging for follow-up

aababilov replied:
Added (in `TableLoaderGenerator.java`).

f8full commented on core/src/main/java/org/mobilitydata/gtfsvalidator/notice/UnknownFileNotice.java:
```diff
@@ -18,8 +18,11 @@
+/**
+ * A file is unknown.
+ */
+public class UnknownFileNotice extends Notice {
```
👍

barbeau commented:
@lionel-nj Do you know why your last commit shows that the test has passed, but the PR is still showing a Circle CI failure (for `ci/circleci: test`?

barbeau commented:
@lionel-nj Oh, nevermind, it shows as passed now. My browser or GitHub must have cached something, and when I posted the above commit it shows green now 👍
